src/vision/status_quo/alan_thinks_he_needs_async_locks.md

```diff
@@ -0,0 +1,127 @@
+      let routes = self.state.lock().unwrap();
+      let route = routes.get(key).unwrap_or_else(crate::error::MissingRoute)?.clone();
+      drop(routes);
+      Ok(self.client.perform_request(route, request).await?)
+  }
+}
+```
+
+This compiles fine and works in testing! After shipping to production, they notice a large increase in throughput. It seems their change made a big difference. Alan is really excited about Rust, and wants to write more!
+
+Alan continues his journey of learning even more about async Rust. After some enlightening talks at the latest RustConf, he decides to revisit the code that he and Barbara wrote together. He asks himself, is using an *async* lock the right thing to do? This lock should only be held for a very short amount of time. Yielding to the runtime is likely more expensive than just synchronously locking. But he remembers vaguely hearing that you should never use blocking code in async code as this will block the entire async executor from being able to make progress, so he doubts his intuition.
+
+After chatting with Barbara, who encourages him to benchmark and measure, he decides to switch back to synchronous locks. 
+
+Unfortunately, switching back to synchronous locks brings back the old compiler error message about his future not being `Send`. Alan is confused as he's dropping the mutex guard before it ever crosses an await point.
+
+Confused Alan goes to Barbara for advice. She is also confused, and it takes several minutes of exploration before she comes to a solution that works: wrapping the mutex access in a block and implicitly dropping the mutex.
+
+```rust
+impl Proxy {
+  async fn handle(&self, key: String, request: Request) -> crate::Result<Response> {
+      let route = {
+        let routes = self.state.lock().unwrap();
+        routes.get(key).unwrap_or_else(crate::error::MissingRoute)?.clone()
+      };
+      Ok(self.client.perform_request(route, request).await?)
+  }
+}
+```
+
+Barbara mentions she's unsure why explicitly dropping the mutex guard did not work, but they're both happy that the code compiles. In fact it seems to have improved the performance of the service when its under extreme load. Alan's intuition was right!
+
+In the end, Barbara decides to write a blog post about how blocking in async code isn't always such a bad idea. 
+
+## 🤔 Frequently Asked Questions
+
+### **What are the morals of the story?**
+    * Locks can be quite common in async code as many tasks might need to mutate some shared state.
+    * Error messages can be fairly good, but they still require a decent understanding of Rust (e.g., `Send`, `MutexGuard`, drop semantics) to fully understand what's going on.
+    * This can lead to needing to use certain patterns (like dropping mutex guards early) in order to get code working.
+    * The advice to never block in async code is not always true: if blocking is short enough, is it even blocking at all?
+### **What are the sources for this story?**
```

Comment:
Maybe cite [this chapter](https://rust-lang.github.io/async-book/07_workarounds/03_send_approximation.html) in the rust book? There's an issue, too, but I can't remember the name of it.

Reply:
Cite https://github.com/rust-lang/rust/issues/57478 ?

Reply:
Done!